gninasrc/lib/bfgs.cu

```diff
@@ -2,21 +2,53 @@
 #include "conf_gpu.h"
 #include "matrix.h"
 #include "bfgs.h"
+#include "device_buffer.h"
 
 #include <cuda_runtime.h>
 
-thread_local float_buffer buffer;
+static __device__
+fl new_lambdamin_val(fl p_val, fl x_val, fl cur_val)
+{
+    // return fabsf(p.values[i]) / fmaxf(fabsf(x.values[i]),
+    //                                   1.0f);
+    fl v = fabsf(p_val / fmaxf(fabsf(x_val), 1.0f));
+    return fmaxf(v, cur_val);
+}
 
-__device__ fl compute_lambdamin(const change_gpu& p, const conf_gpu& x, sz n)
+__device__ fl compute_lambdamin(const change_gpu& p, const conf_gpu& x, sz n, const gpu_data& d)
 
 {
+    const tree_gpu& tree = *d.treegpu;
     fl test = 0;
-	for (sz i = 0; i < n; i++)
+	for (sz i = 0; i < tree.num_nodes; i++)
 	{
-		fl temp = fabsf(p.values[i]) / fmaxf(fabsf(x.values[i]),
-                                                    1.0f);
-		if (temp > test)
-			test = temp;
+        if(i < tree.nlig_roots){
+            float* conf_start = &x.values[i * 7];
+            float* change_start = &p.values[i * 6];
+
+            for(sz i = 0; i < 3; i++)
+                test = new_lambdamin_val(change_start[i], conf_start[i], test);
+
+            qt orientation(conf_start[3],
+                           conf_start[4],
+                           conf_start[5],
+                           conf_start[6]);
+            // qt *orientation = (qt *)(void*) &conf_start[3];
+            vec angle = quaternion_to_angle(orientation);
+            for(sz i = 0; i < 3; i++)
+                test = new_lambdamin_val(change_start[3 + i], angle[i], test);
+
+        } else {
+            float* conf_start = &x.values[i + tree.nlig_roots * 6];
+            float* change_start = &p.values[i + tree.nlig_roots * 5];
```

Comment:
slightly awkward to take address and then immediately dereference instead of go straight to float